apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/ClientPool.cs

```diff
@@ -87,50 +92,137 @@ public override int GetHashCode()
             }
         }
 
-        private class ClientPoolEntry
+        private class CredentialClientPool
         {
-            private readonly ClientPoolKey _key;
-            private SpannerClient _client;
-            private volatile Task<SpannerClient> _creationTask;
+            private readonly ClientCredentialKey _key;
+            private SortedList<SpannerClientCreator, SpannerClientCreator> _multiClientPool = 
+                new SortedList<SpannerClientCreator, SpannerClientCreator>();
             private readonly object _sync = new object();
 
-            public ClientPoolEntry(ClientPoolKey key) => _key = key;
 
-            public async Task<SpannerClient> AcquireClientFromEntryAsync()
+            public CredentialClientPool(ClientCredentialKey key) => _key = key;
+
+            public Task<SpannerClient> AcquireClientAsync()
             {
-                if (_client == null)
+                Task<SpannerClient> result;
+
+                lock (_sync)
                 {
-                    Logger.Debug(() => "Creating a new SpannerClient.");
-                    var endpoint = _key.Endpoint ?? SpannerClient.DefaultEndpoint;
-                    if (_key.Credential != null)
+                    //first ensure that the pool is of the correct size.
+                    while (_multiClientPool.Count > SpannerOptions.Instance.MaximumGrpcChannels)
                     {
-                        lock (_sync)
-                        {
-                            if (_client == null)
-                            {
-                                var channel = new Channel(
-                                    endpoint.Host,
-                                    endpoint.Port,
-                                    _key.Credential.ToChannelCredentials());
-                                _client = SpannerClient.Create(channel);
-                            }
-                        }
+                        _multiClientPool.RemoveAt(_multiClientPool.Count - 1);
                     }
-                    else
+                    while (_multiClientPool.Count < SpannerOptions.Instance.MaximumGrpcChannels)
                     {
-                        lock (_sync)
-                        {
-                            if (_creationTask == null || _creationTask.IsFaulted)
-                            {
-                                _creationTask = SpannerClient.CreateAsync(endpoint);
-                            }
-                        }
-                        // await needs to be done outside of the monitor.
-                        _client = await _creationTask.ConfigureAwait(false);
+                        var newEntry = new SpannerClientCreator(_key);
+                        _multiClientPool.Add(newEntry, newEntry);
                     }
+
+                    //now grab the first item in the sorted list, increment refcnt, re-sort and return.
```

Comment:
decided that the latest implementation has the least risk of state getting out of sync.  It might have been slightly better for items to remember the prioritykey that they last used to be inserted into the list -- but with my latest commit, the items only need to fire an event when they know their priority has changed and do not need to remember past state at all.